# NGS_Cross_Contamination

## Possible Errors:
# TODO: Set the errors in the correct order
1) parameters.py: Error 1: Entered file has incorrect extension.
2) parameters.py: Error 2: Entered file has incorrect extension. 
3) parameters.py: Error 3: Entered indexing value is incorrect.
4) parameters.py: Error 4: Entered analysis value is incorrect.
5) parameters.py: Error 5: Entered maximum nucleotide difference between barcodes is not numeric.
6) parameters.py: Error 6: Entered maximum nucleotide difference between sequences is not numeric.
7) parameters.py: Error 7: Entered sequencer value is incorrect.
8) parameters.py: Error 8: Entered output directory path does not exist.
9) parameters.py: Error 9: Missing one of the parameters. Make sure you have entered all parameters in parameters.txt.
10) file_readers.py: Error 10: Fasta file format is incorrect. No headers found.
11) file_readers.py: Error 11: Entered fastq file has not been found.
12) check_barcodes.py: Error 12: Barcodes have not been found at the end of the headers in the fastq file.
13) file_readers.py: Error 13: Entered barcode file has not been found.
14) file_readers.py: Error 14: Barcode file format incorrect.
15) file_readers.py: Error 15: Barcode file format incorrect.
16) parameters.py: Error 16: Entered heatmap contamination percentage is not a number.
17) parameters.py: Error 17:  Entered spike-ins trimming value is not numeric.
18) parameters.py: Error 18: Entered spike-ins i5 trimming value is not numeric.
19) parameters.py: Error 19: Entered spike-ins i7 trimming value is not numeric.
20) Parameters.py: Error 20: Entered spike-ins order value is not numeric.