# NGS_Cross_Contamination

## Code manual:
### TODO: Add a manual on how to use the program. 

## List of errors that can occur when running the code:
1) parameters.py - Error 1: Entered fastq file has incorrect file extension.
2) parameters.py - Error 2: Entered barcode (+ spike) file has incorrect file extension.
3) parameters.py - Error 3: Entered indexing value is incorrect.
4) parameters.py - Error 4: Entered analysis combination value is incorrect.
5) parameters.py - Error 5: Entered analysis type value is incorrect.
6) parameters.py - Error 6: Barcode nucleotide difference parameter is incorrect.
7) parameters.py - Error 7: Spike-in nucleotide difference parameter is incorrect.
8) parameters.py - Error 8: Entered sequencer value is incorrect.
9) parameters.py - Error 9: Entered output directory path does not exist.
10) parameters.py - Error 10: Entered heatmap contamination percentage is incorrect.
11) parameters.py - Error 11: Entered i7 barcode length is incorrect.
12) parameters.py - Error 12: Entered spike-ins left trimming value is incorrect.
13) parameters.py - Error 13: Entered spike-ins right trimming value is incorrect.
14) parameters.py - Error 14: Not all parameters have been entered in settings.py.
15) file_readers.py - Error 15: Fasta file format is incorrect. No headers found.
16) file_readers.py - Error 16: Entered fastq file has not been found.
17) file_readers.py - Error 17: Entered barcode file has not been found.
18) file_readers.py - Error 18: Barcode file format incorrect.
19) file_readers.py - Error 19: Fasta file format is incorrect. No headers or sequences found.
20) file_readers.py - Error 20: Entered fastq file has not been found.